relay-client/src/client.rs

```diff
@@ -0,0 +1,749 @@
+            .ok_or_else(|| eyre::eyre!("Client not connected"))?;
+        command_tx.send(Command::Reconnect).await?;
+        Ok(())
+    }
+
+    pub async fn graceful_shutdown(
+        &mut self,
+        wait_for_pending_messages: Duration,
+        wait_for_shutdown: Duration,
+    ) {
+        // Let's wait for all acks to be received
+        if self.has_pending_messages().await.map_or(false, |n| n > 0) {
+            tracing::info!(
+                "Giving {}ms for pending messages to be acked",
+                wait_for_pending_messages.as_millis()
+            );
+            tokio::time::sleep(wait_for_pending_messages).await;
+        }
+        // If there are still pending messages, we retry to send them
+        if self.has_pending_messages().await.map_or(false, |n| n > 0) {
+            tracing::info!("There are still pending messages, replaying...");
+            if let Ok(()) = self.replay_pending_messages().await {
+                tokio::time::sleep(wait_for_pending_messages).await;
+            }
+        }
+
+        // Eventually, there not much more we can do, so we shutdown the client
+        self.shutdown();
+
+        if let Some(shutdown_completed) = self.shutdown_completed.take() {
+            match tokio::time::timeout(wait_for_shutdown, shutdown_completed).await {
+                Ok(_) => tracing::info!("Shutdown completed successfully."),
+                Err(_) => tracing::warn!("Timed out waiting for shutdown to complete."),
+            }
+        }
+    }
+
+    /// Shutdown the client
+    pub fn shutdown(&mut self) {
+        tracing::info!("Shutting down requested");
+        if let Some(token) = self.shutdown_token.take() {
+            token.cancel();
+        }
+    }
+
+    pub async fn wait_for_msg_while_spamming<
+        T: PayloadMatcher,
+        S: IntoPayload + std::clone::Clone,
+    >(
+        &mut self,
+        wait: Duration,
+        spam: S,
+        spam_every: Duration,
+    ) -> Result<T::Output> {
+        let start_time = tokio::time::Instant::now();
+        let mut spam_time = tokio::time::Instant::now();
+        loop {
+            if let Some(payload) = self.check_for_msg::<T>().await {
+                return Ok(payload);
+            }
+
+            if spam_time.elapsed() >= spam_every {
+                let _ = self.send(spam.clone()).await;
+                spam_time = tokio::time::Instant::now();
+            }
+
+            if start_time.elapsed() >= wait {
+                return Err(eyre::eyre!(
+                    "Timeout waiting for payload of type {:?}",
+                    std::any::type_name::<T>()
+                ));
+            }
+            tokio::time::sleep(Duration::from_millis(100)).await;
+        }
+    }
+}
+
+impl Drop for Client {
+    fn drop(&mut self) {
+        self.shutdown();
+    }
+}
+
+struct PollerAgent<'a> {
+    config: &'a Config,
+    pending_messages: BTreeMap<u64, (RelayConnectRequest, Option<oneshot::Sender<()>>)>,
```

Comment:
suggestion: newtype the u64 to improve documentation and type safety like this:

```rust
#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Clone, Copy, derive_more::Deref, derive_more::Into)]
struct AckNum(u64);
```

Reply:
Will do in PR #342